From stable-bounces@linux.kernel.org Wed Nov  2 15:36:17 2005
Date: Thu, 3 Nov 2005 07:55:38 +1100
To: Arnaldo Carvalho de Melo <acme@ghostprotocols.net>, netdev@vger.kernel.org, stable@kernel.org
Message-ID: <20051102205538.GA24276@gondor.apana.org.au>
Content-Disposition: inline
From: Herbert Xu <herbert@gondor.apana.org.au>
Cc: phillips@istop.com
Subject: [NET] Fix zero-size datagram reception

The recent rewrite of skb_copy_datagram_iovec broke the reception of
zero-size datagrams.  This patch fixes it.

Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 net/core/datagram.c |    4 ++++
 1 file changed, 4 insertions(+)

--- linux-2.6.14.1.orig/net/core/datagram.c
+++ linux-2.6.14.1/net/core/datagram.c
@@ -213,6 +213,10 @@ int skb_copy_datagram_iovec(const struct
 {
 	int i, err, fraglen, end = 0;
 	struct sk_buff *next = skb_shinfo(skb)->frag_list;
+
+	if (!len)
+		return 0;
+
 next_skb:
 	fraglen = skb_headlen(skb);
 	i = -1;
